protocols/noise/src/lib.rs

```diff
@@ -48,195 +48,275 @@
     }
-}
-
-impl<C: Protocol<C> + Zeroize> NoiseConfig<IK, C> {
-    /// Create a new `NoiseConfig` for the IK handshake pattern (recipient side).
-    pub fn ik_listener(keys: Keypair<C>) -> Self {
-        NoiseConfig {
-            keys,
-            params: C::params_ik(),
-            remote: (),
-            _marker: std::marker::PhantomData
+    fn get_remote_static(&self) -> Result<[u8; 32], NoiseError> {
+        match self {
+            NoiseSession::ik(a) => Ok(a.get_remote_static_public_key().as_bytes()),
+            NoiseSession::ix(a) => Ok(a.get_remote_static_public_key().as_bytes()),
+            NoiseSession::xx(a) => Ok(a.get_remote_static_public_key().as_bytes()),
+            _ => Err(NoiseError::__Nonexhaustive),
         }
     }
 }
 
-impl<C: Protocol<C> + Zeroize> NoiseConfig<IK, C, PublicKey<C>> {
-    /// Create a new `NoiseConfig` for the IK handshake pattern (initiator side).
-    pub fn ik_dialer(keys: Keypair<C>, remote: PublicKey<C>) -> Self {
-        NoiseConfig {
-            keys,
-            params: C::params_ik(),
-            remote,
-            _marker: std::marker::PhantomData
-        }
+// Handshake pattern IX /////////////////////////////////////////////////////
+#[derive(Clone)]
+pub struct IX(Keypair);
+
+impl IX {
+    /// Create a new `NoiseConfig` for the IX handshake pattern.
+    pub fn new(k: Keypair) -> IX {
+        IX(k)
     }
 }
 
-// Handshake pattern IX /////////////////////////////////////////////////////
+impl UpgradeInfo for IX {
+    type Info = &'static [u8];
+    type InfoIter = std::iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        std::iter::once(b"/noise/ix/25519/chachapoly/blake2s/0.1.0")
+    }
+}
 
-impl<T, C> InboundUpgrade<T> for NoiseConfig<IX, C>
+impl<T> InboundUpgrade<T> for IX
 where
     T: AsyncRead + AsyncWrite,
-    NoiseConfig<IX, C>: UpgradeInfo,
-    C: Protocol<C> + AsRef<[u8]> + Zeroize
+    IX: UpgradeInfo,
 {
-    type Output = (PublicKey<C>, NoiseOutput<Negotiated<T>>);
+    type Output = ([u8; 32], NoiseOutput<Negotiated<T>>);
     type Error = NoiseError;
-    type Future = rt1::NoiseInboundFuture<Negotiated<T>, C>;
+    type Future = rt1::NoiseInboundFuture<Negotiated<T>>;
 
     fn upgrade_inbound(self, socket: Negotiated<T>, _: Self::Info) -> Self::Future {
-        let session = self.params.into_builder()
-            .local_private_key(self.keys.secret().as_ref())
-            .build_responder()
-            .map_err(NoiseError::from);
-        rt1::NoiseInboundFuture::new(socket, session)
+        if let Ok(prologue) = Message::from_bytes(b"Noise_IX_25519_ChaChaPoly_Blake2s") {
+            let session = NoiseSession::ix(noisesession_ix::NoiseSession::init_session(
+                false,
+                prologue,
+                self.0,
+                PublicKey::empty(),
+            ));
+            return rt1::NoiseInboundFuture::new(socket, session);
+        }
+        panic!("Should not panic");
```

Comment:
I think this should be removed, and you should have `Message::from_bytes(...).unwrap()` instead.